testtime/testtime.go

```diff
@@ -0,0 +1,213 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package testtime provides a mocked version of time.Timer for use in tests.
+package testtime
+
+import (
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/timeutil"
+)
+
+// TestTimer is a mocked version of time.Timer for which the passage of time or
+// the direct expiration of the timer is controlled manually.
+//
+// TestTimer implements timeutil.Timer.
+//
+// TestTimer also provides methods to introspect whether the timer is active or
+// how many times it has fired.
+type TestTimer struct {
+	lock       sync.Mutex
+	currTime   time.Time
```

Comment:
it seems we could keep a elapsed Duration instead of this?

Reply:
Good idea, I think elapsed duration reflects what we're trying to represent better than a timestamp based on the creation time of the timer.